benchmark/kmeans.py

```diff
@@ -0,0 +1,193 @@
+                algorithm="lloyd",
+            )
+
+            sklearn.cluster._kmeans._kmeans_single_lloyd = kmeans_fn
+            estimator = KMeans(**est_kwargs)
+
+            print(f"Running {name} with parameters max_iter={max_iter} tol={tol} ...")
+
+            # dry run, to be fair for JIT implementations
+            KMeans(**est_kwargs).set_params(max_iter=2).fit(
+                X, sample_weight=sample_weight
+            )
+
+            t0 = perf_counter()
+            estimator.fit(X, sample_weight=sample_weight)
+            t1 = perf_counter()
+
+            self._check_same_fit(
+                estimator, name, max_iter, assert_allclose=run_consistency_test
+            )
+            print(f"Running {name} ... done in {t1 - t0}\n")
+
+        finally:
+            sklearn.cluster._kmeans._kmeans_single_lloyd = VANILLA_SKLEARN_LLOYD
+
+    def _check_same_fit(self, estimator, name, max_iter, assert_allclose):
+        text = (
+            "It is expected for all iterators in the benchmark to run the same "
+            "number of iterations and return the same results."
+        )
+        n_iter = estimator.n_iter_
+        if n_iter != max_iter:
+            raise RuntimeError(
+                f"The estimator {name} only ran {n_iter} iterations instead of "
+                f"max_iter={max_iter} iterations. " + text
+            )
+
+        if not assert_allclose:
+            return
+
+        if self.results is None:
+            self.results = dict(
+                labels=estimator.labels_,
+                cluster_centers=estimator.cluster_centers_,
+                inertia=estimator.inertia_,
+            )
+        else:
+            assert_array_equal(self.results["labels"], estimator.labels_, err_msg=text)
+            assert_allclose(
+                self.results["cluster_centers"],
+                estimator.cluster_centers_,
+                err_msg=text,
+            )
+            assert_allclose(self.results["inertia"], estimator.inertia_, err_msg=text)
+
+    def _check_kmeans_fn_signature(self, kmeans_fn):
+        fn_signature = inspect.signature(kmeans_fn)
+        if fn_signature != self._VANILLA_SKLEARN_LLOYD_SIGNATURE:
+            raise ValueError(
+                f"The signature of the submitted kmeans_fn is expected to be "
+                f"{self._VANILLA_SKLEARN_LLOYD_SIGNATURE}, but got {fn_signature} ."
+            )
+
+
+if __name__ == "__main__":
+    # TODO: also checks that the results are close
+    # TODO: ensure that effective n_iter is equal for all runs
+
+    from ext_helpers.sklearn import kmeans as sklearn_kmeans
+    from ext_helpers.daal4py import kmeans as daal4py_kmeans
+
+    from sklearn_numba_dpex.kmeans.drivers import LLoydKMeansDriver
+    from sklearn.datasets import fetch_openml
+    from sklearn.preprocessing import MinMaxScaler
+    from sklearnex import config_context
+    from numpy.random import default_rng
+    import numpy as np
```

Comment:
Should those be moved at the top of the file ?

Reply:
I don't think so ? the imports at the top are the one necessary to define the objects what would be seen by other modules, the imports after the `__main__` are only necessary if the file is explicitly run.